sdk/azidentity/client_secret_credential.go

```diff
@@ -54,47 +55,62 @@ func (c *ClientSecretCredential) AuthenticationPolicy(options azcore.Authenticat
 }
 
 type bearerTokenPolicy struct {
-	// take lock when manipulating header/expiresOn fields
-	lock      sync.RWMutex
-	header    string
-	expiresOn time.Time
+	updating  int64                  // atomically set to 1 to indicate a refresh is in progress
+	header    atomic.Value           // string
+	expiresOn atomic.Value           // time.Time
 	creds     azcore.TokenCredential // R/O
 	scopes    []string               // R/O
 }
 
 func newBearerTokenPolicy(creds azcore.TokenCredential, scopes []string) *bearerTokenPolicy {
-	// set the token as expired so first call to Do() refreshes it
-	return &bearerTokenPolicy{creds: creds, scopes: scopes, expiresOn: time.Now().UTC()}
+	bt := bearerTokenPolicy{creds: creds, scopes: scopes}
+	// set initial values to their zero-value
+	bt.header.Store("")
+	bt.expiresOn.Store(time.Time{})
+	return &bt
 }
 
 func (b *bearerTokenPolicy) Do(ctx context.Context, req *azcore.Request) (*azcore.Response, error) {
+	// TODO: is window really constant, what about short-lived tokens?
+	const window = 5 * time.Minute
 	var bt string
-	// take read lock and check if the token has expired
-	b.lock.RLock()
+	// check if the token has expired
 	now := time.Now().UTC()
-	if now.Equal(b.expiresOn) || now.After(b.expiresOn) {
-		// token has expired, take the write lock then check again
-		b.lock.RUnlock()
-		// don't defer Unlock(), we want to release it ASAP
-		b.lock.Lock()
-		if now.Equal(b.expiresOn) || now.After(b.expiresOn) {
-			// token has expired, get a new one and update shared state
+	exp := b.expiresOn.Load().(time.Time)
+	if now.After(exp) {
```

Comment:
I think you want to use the window here; if the expiration is within 5 seconds of NOW

Reply:
The value stored in `expiresOn` has the window subtracted from it (see line 95).

Reply:
I would do it this way. I'd stop the actual expiration time that the service gives back to us in the field. And after getting the value, subtract the windows and then compare to NOW. 

This way, the field always stores the actual value returned from the service which is really useful for debugging - we know what the service returned (not some mutated value). We mutate the value ephemerally just for comparison.

Also, imagine someday, we dynamically adjust the window as the program runs. The expires time is always the actual expires times and we can dynamically adjust the window over time WHEN we do the comparison.

Reply:
Fixed.